contributors/design-proposals/kubelet-server-certificate-bootstrap-rotation.md

```diff
@@ -0,0 +1,97 @@
+# Kubelet Server Certificate Bootstrap & Rotation
+
+## Overview
+
+Currently, when a kubelet first starts, it generates a self signed
+certificate/key pair that is used for accepting incoming TLS connections. This
+proposal covers a process for generating a key locally and then issuing a
+Certificate Signing Request to the cluster API server to get an associated
+certificate signed by the cluster Certificate Authority. Also, as certificates
+approach expiration, the same mechanism will be used to request an updated
+certificate.
+
+## Use Cases
+
+1. Kubelet starts for the first time on a new node.
+2. Kubelet certificate approaches expiration.
+
+## Proposed Design
+
+### During Kubelet Boot Sequence
+
+1. Look on disk for an existing cert/key pair.
+1. If there is an existing cert/key pair, load them.
+1. If there is no existing cert/key pair, at the point where the kubelet
+   currently generates a key and certificate, the kubelet will:
+    - generate a key
+    - create a certificate signing request
+    - request a signed certificate from the API server
+    - wait for a response
+    - Store the new cert/key pairs in the directory specified by `--cert-dir`,
+      or the default value if it is unspecified.
+
+### As Expiration Approaches
+
+Rotating a Kubelet server certificate will work by generating a new private key,
+issuing a new Certificate Signing Request to the API Server, safely updating the
+cert/key pair on disk, begin using the new cert/key pair.
+ 
+1. Store the new cert/key pairs in the directory specified by `--cert-dir`, or the
+   default value if it is unspecified. This will allow the kubelet to have a
+   place for storing the multiple cert/key pairs that it might have available at
+   any given moment (because of a rotation in progress).
+    - When cert/key data is directly embedded in the kubelet config file
+      (client-certificate-data, client-key-data)
+        - The values be used only if there are no updated cert/key pairs due to
+          rotation
+    - When cert/key files are specified in kubeconfig, or with the
+      `--tls-cert-file` and `--tls-private-key-file`:
+        - These values be used only if there are no updated cert/key pairs due
+          to rotation
+1. There will be a feature gate to enable certificate bootstrapping and
+   rotation, `RotateKubeletServerCertificate`
+1. Centralize certificate access within the kubelet code to the
+   CertificateManager. The CertificateManager will be responsible for:
+    - Providing the correct certificate to use for establishing TLS connections.
+    - Generating new private keys and requesting new certificates when the
+      current certificate approaches expiry.
+    - Since certificates can rotate at any time, all other parts of the kubelet
+      should ask the CertificateManager for the correct certificate each time a
+      certificate is used. No certificate caching except by the
```

Comment:
Ah I see that #41912 actually added the GetClientCertificate field to the TLS config.

Does it work for the cache to ignore configs that set this new field? I'm under the impression that the kubelet is a lot better about keeping a centralized client instead of constantly creating new ones like kubectl.